### PR TITLE
Calibration Methods

### PR DESCRIPTION
**Pull request by DolanTheMFWizard**
*Issue #, if available:*

*Description of changes:*
Made into PR, because Dylan's last day. Further things to explore if AG team is interested in more calibration methods to apply to AG.

A lot of these look really bad. Dirichlet calibrate is the only one with some promise, but in some cases severely mis-calibrated the model. More exploration is probably required.

pulled off of https://pypi.org/project/dirichletcal/. Had to clone and  change repo as it uses a really outdated version of scikit-learn not compatible with AG.

All methods retrofitted from https://github.com/dirichletcal/dirichlet_python
 More information on methods:
[Beyond temperature scaling: Obtaining well-calibrated multiclass probabilities with Dirichlet calibration](https://arxiv.org/abs/1910.12656)
[On Calibration of Modern Neural Networks](https://arxiv.org/abs/1706.04599)

`tempscaling_og.py` is the one with 65% WR over vanilla ag. `tempscaling.py` is the one that did worse. Something very suspicious is going on. Temperature scaling should not do worse on all datasets if implemented properly. It's a single parameter so it can't overfit and it's trained to minimize NLL on validation set, so if the test set is not covariate shifted it should improve test scores.

`/opt/anaconda3/envs/AG/bin/python "/Users/yundylan/Documents/awslabs GitHub/autogluon-utils/autogluon_utils/benchmarking/evaluation/runners/run_evaluation_openml.py"

binary        2002
multiclass    1352
Name: problem_type, dtype: int64
helena                 57.829807
micro-mass             12.866110
yeast                   7.855258
segment                 7.261750
robert                  6.771462
                         ...    
guillermo              -3.837350
sf-police-incidents    -4.398973
churn                  -5.629881
connect-4              -7.989737
albert                 -8.247718
Length: 64, dtype: float64
Win 15 : Conflation_results_automlbenchmark_11_17_conformal
Win 12 : Conflation_dirichlet_results_automlbenchmark_11_17_conformal
Tie 37
Higgs                               108.706301
volkert                              89.800283
sf-police-incidents                  89.713211
airlines                             78.488953
albert                               65.554453
                                       ...    
kc1                                   3.002288
qsar-biodeg                           2.988282
blood-transfusion-service-center      1.877825
arcene                                1.000883
Satellite                             0.583689
Length: 64, dtype: float64
Win 61 : Conflation_results_automlbenchmark_11_17_conformal
Win 0 : Conflation_matrix_results_automlbenchmark_11_17_conformal
Tie 3
helena                 270.394174
sf-police-incidents    186.832316
Higgs                  128.161741
airlines               123.525212
volkert                110.922593
                          ...    
PhishingWebsites         5.219960
qsar-biodeg              4.357679
porto-seguro             2.541253
Satellite                1.881389
APSFailure               1.506867
Length: 64, dtype: float64
Win 62 : Conflation_results_automlbenchmark_11_17_conformal
Win 0 : Conflation_temp_results_automlbenchmark_11_17_conformal
Tie 2
helena                              193.486081
porto-seguro                        135.655173
Higgs                                97.801956
amazon-commerce-reviews              89.081219
airlines                             79.339833
                                       ...    
kc1                                   2.908829
qsar-biodeg                           2.532634
blood-transfusion-service-center      1.734919
arcene                                0.716010
Satellite                             0.429526
Length: 63, dtype: float64
Win 60 : Conflation_results_automlbenchmark_11_17_conformal
Win 0 : Conflation_vector_results_automlbenchmark_11_17_conformal
Tie 3
Z-statistic comparisons:
                     Conflation_dirichlet_results_automlbenchmark_11_17_conformal  Conflation_matrix_results_automlbenchmark_11_17_conformal  Conflation_temp_results_automlbenchmark_11_17_conformal  Conflation_vector_results_automlbenchmark_11_17_conformal        min
helena                                                       57.829807                                                     12.566098                                                 270.394174                                               193.486081          12.566098
micro-mass                                                   12.866110                                                     11.894672                                                  15.364276                                                13.437179          11.894672
yeast                                                         7.855258                                                      8.226539                                                  15.607098                                                11.168712           7.855258
segment                                                       7.261750                                                     18.636366                                                  28.493398                                                18.901766           7.261750
robert                                                        6.771462                                                     22.279653                                                  37.552989                                                28.607503           6.771462
...                                                                ...                                                           ...                                                        ...                                                      ...                ...
guillermo                                                    -3.837350                                                     15.849377                                                  24.653941                                                15.946270          -3.837350
sf-police-incidents                                          -4.398973                                                     89.713211                                                 186.832316                                                      NaN          -4.398973
churn                                                        -5.629881                                                     10.817949                                                  17.595558                                                10.557489          -5.629881
connect-4                                                    -7.989737                                                     32.070439                                                  56.432860                                                44.236751          -7.989737
albert                                                       -8.247718                                                     65.554453                                                  90.704020                                                64.190454          -8.247718

[65 rows x 5 columns]
num_datasets: 705
num_folds: 1
errors: 171
################################################
framework: Conflation_results_automlbenchmark_11_17_conformal
datasets_framework_errors: ['KDDCup09-Upselling_1', 'KDDCup09-Upselling_3', 'KDDCup09-Upselling_4', 'KDDCup09-Upselling_5', 'KDDCup09-Upselling_7', 'numerai28_6_4']
datasets_framework_errors_count: 6
framework_fold_errors: 6
################################################
################################################
framework: Conflation_dirichlet_results_automlbenchmark_11_17_conformal
datasets_framework_errors: ['APSFailure_4', 'KDDCup09-Upselling_1', 'KDDCup09-Upselling_3', 'KDDCup09-Upselling_4', 'KDDCup09-Upselling_5', 'KDDCup09-Upselling_7', 'KDDCup99_0', 'KDDCup99_1', 'KDDCup99_2', 'KDDCup99_3', 'KDDCup99_4', 'KDDCup99_5', 'KDDCup99_6', 'KDDCup99_7', 'KDDCup99_8', 'KDDCup99_9', 'dionis_0', 'dionis_1', 'dionis_2', 'dionis_3', 'dionis_4', 'dionis_5', 'dionis_6', 'dionis_7', 'dionis_8', 'dionis_9', 'shuttle_0', 'shuttle_1', 'shuttle_2', 'shuttle_3', 'shuttle_4', 'shuttle_5', 'shuttle_6', 'shuttle_7', 'shuttle_8', 'shuttle_9', 'wine-quality-white_3', 'wine-quality-white_4', 'wine-quality-white_5', 'wine-quality-white_6', 'wine-quality-white_7', 'wine-quality-white_8', 'wine-quality-white_9']
datasets_framework_errors_count: 43
framework_fold_errors: 43
################################################
################################################
framework: Conflation_matrix_results_automlbenchmark_11_17_conformal
datasets_framework_errors: ['KDDCup09-Upselling_1', 'KDDCup09-Upselling_7', 'KDDCup99_0', 'KDDCup99_1', 'KDDCup99_2', 'KDDCup99_3', 'KDDCup99_4', 'KDDCup99_5', 'KDDCup99_6', 'KDDCup99_7', 'KDDCup99_8', 'KDDCup99_9', 'dionis_0', 'dionis_1', 'dionis_2', 'dionis_3', 'dionis_4', 'dionis_5', 'dionis_6', 'dionis_7', 'dionis_8', 'dionis_9', 'porto-seguro_7', 'shuttle_0', 'shuttle_1', 'shuttle_2', 'shuttle_3', 'shuttle_4', 'shuttle_5', 'shuttle_6', 'shuttle_7', 'shuttle_8', 'shuttle_9', 'wine-quality-white_3', 'wine-quality-white_4', 'wine-quality-white_5', 'wine-quality-white_6', 'wine-quality-white_7', 'wine-quality-white_8', 'wine-quality-white_9']
datasets_framework_errors_count: 40
framework_fold_errors: 40
################################################
################################################
framework: Conflation_temp_results_automlbenchmark_11_17_conformal
datasets_framework_errors: ['KDDCup09-Upselling_3', 'KDDCup99_0', 'KDDCup99_1', 'KDDCup99_2', 'KDDCup99_3', 'KDDCup99_4', 'KDDCup99_5', 'KDDCup99_6', 'KDDCup99_7', 'KDDCup99_8', 'KDDCup99_9', 'arcene_7', 'dionis_0', 'dionis_1', 'dionis_2', 'dionis_3', 'dionis_4', 'dionis_5', 'dionis_6', 'dionis_7', 'dionis_8', 'dionis_9', 'shuttle_0', 'shuttle_1', 'shuttle_2', 'shuttle_3', 'shuttle_4', 'shuttle_5', 'shuttle_6', 'shuttle_7', 'shuttle_8', 'shuttle_9', 'wine-quality-white_3', 'wine-quality-white_4', 'wine-quality-white_5', 'wine-quality-white_6', 'wine-quality-white_7', 'wine-quality-white_8', 'wine-quality-white_9']
datasets_framework_errors_count: 39
framework_fold_errors: 39
################################################
################################################
framework: Conflation_vector_results_automlbenchmark_11_17_conformal
datasets_framework_errors: ['Bioresponse_4', 'KDDCup09-Upselling_1', 'KDDCup09-Upselling_3', 'KDDCup09-Upselling_4', 'KDDCup09-Upselling_5', 'KDDCup99_0', 'KDDCup99_1', 'KDDCup99_2', 'KDDCup99_3', 'KDDCup99_4', 'KDDCup99_5', 'KDDCup99_6', 'KDDCup99_7', 'KDDCup99_8', 'KDDCup99_9', 'dionis_0', 'dionis_1', 'dionis_2', 'dionis_3', 'dionis_4', 'dionis_5', 'dionis_6', 'dionis_7', 'dionis_8', 'dionis_9', 'sf-police-incidents_7', 'shuttle_0', 'shuttle_1', 'shuttle_2', 'shuttle_3', 'shuttle_4', 'shuttle_5', 'shuttle_6', 'shuttle_7', 'shuttle_8', 'shuttle_9', 'wine-quality-white_3', 'wine-quality-white_4', 'wine-quality-white_5', 'wine-quality-white_6', 'wine-quality-white_7', 'wine-quality-white_8', 'wine-quality-white_9']
datasets_framework_errors_count: 43
framework_fold_errors: 43
################################################
0.038035242299606344
0.019391889480048812
0.17809096487748796
0.0024488585123725918
0.25897345381903814
0.0022539886352489916
0.1845496600337805
0.0023821666284783334
################################################
all VS all
	All datasets regardless of failures
valid_tasks: 705
                                           framework  time_train_s  metric_error  time_infer_s  bestdiff  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0  Conflation_results_automlbenchmark_11_17_confo...    846.014306      0.384117     12.508870  0.020410       0.049146               1.003702               1.091011  1.525036           380           288            18            13            6
1  Conflation_dirichlet_results_automlbenchmark_1...    810.884894      0.410903      5.811027  0.040118       0.094487               1.313372               1.441703  1.725831           286           317            22            37           43
2  Conflation_matrix_results_automlbenchmark_11_1...    785.665714      0.458086      5.973233  0.188904       0.501316               1.124752               1.437089  3.352632            11            26           287           341           40
3  Conflation_vector_results_automlbenchmark_11_1...    774.963444      0.471002      5.602568  0.195718       0.522387               1.130701               1.438007  3.486405             7            28           205           422           43
4  Conflation_temp_results_automlbenchmark_11_17_...    790.230931      0.531865      6.315916  0.267734       0.951553               1.105533               1.428535  4.830330            10            11            13           632           39

################################################
all VS all
	Only datasets where all frameworks succeeded
valid_tasks: 657
                                           framework  time_train_s  metric_error  time_infer_s  bestdiff  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0  Conflation_results_automlbenchmark_11_17_confo...    762.082801      0.393182      5.483866  0.021567       0.051347               1.003915               1.096527  1.554033           340           287            17            13            6
1  Conflation_dirichlet_results_automlbenchmark_1...    807.525114      0.410250      5.583562  0.040185       0.092100               1.315335               1.443960  1.719178           285           315            22            35           43
2  Conflation_matrix_results_automlbenchmark_11_1...    773.329680      0.459527      5.469406  0.190341       0.504699               1.126046               1.441498  3.369102             8            24           285           340           40
3  Conflation_vector_results_automlbenchmark_11_1...    771.824353      0.471862      5.503957  0.196350       0.524276               1.131440               1.441167  3.497717             6            26           203           422           43
4  Conflation_temp_results_automlbenchmark_11_17_...    771.209589      0.535424      5.684323  0.270214       0.955376               1.106185               1.433038  4.859970             9             7            12           629           39

################################################
Conflation_results_automlbenchmark_11_17_conformal VS all
                                           framework   winrate    >    <    =  % Less Avg. Errors  time_train_s  metric_error  time_infer_s  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0  Conflation_results_automlbenchmark_11_17_confo...  0.500000    0    0  699            0.000000    762.082801      0.393182      5.483866       0.051347               1.003915               1.096527  1.554033           380           288            18            13            6
1  Conflation_dirichlet_results_automlbenchmark_1...  0.449319  296  363    2           -1.864335    807.525114      0.410250      5.583562       0.092100               1.315335               1.443960  1.719178           286           317            22            37           43
2  Conflation_matrix_results_automlbenchmark_11_1...  0.037065   24  636    1          -17.564211    773.329680      0.459527      5.469406       0.504699               1.126046               1.441498  3.369102            11            26           287           341           40
3  Conflation_vector_results_automlbenchmark_11_1...  0.037879   25  635    0          -18.216749    771.824353      0.471862      5.503957       0.524276               1.131440               1.441167  3.497717             7            28           205           422           43
4  Conflation_temp_results_automlbenchmark_11_17_...  0.031014   20  640    1          -25.671947    771.209589      0.535424      5.684323       0.955376               1.106185               1.433038  4.859970            10            11            13           632           39

Process finished with exit code 0
`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Closing this PR for now as initial results were not showing improvement and code requires a fair amount of cleanup prior to being merge-able. Feel free to re-open this PR if progress is continuing towards making it merge-able.
